Modules/rat_command.py

```diff
@@ -76,18 +70,82 @@ class NameCollisionException(CommandException):
+    if message.strip() == "":
+        return
+
+    words, words_eol = _split_message(message.lstrip(prefix))
+    if message.startswith(prefix):
+        if words[0].casefold() in _registered_commands.keys():
+            # A regular command
+            command_fun = _registered_commands[words[0].casefold()]
+            extra_args = ()
+            log.debug(f"Regular command {words[0]} invoked.")
+        else:
+            # Might be a regular rule
+            command_fun, extra_args = _get_rule(words, words_eol, _rules)
+            if command_fun:
+                log.debug(f"Rule {getattr(command_fun, '__name__', '')} matching {words[0]} found.")
+            else:
+                log.warning(f"Could not find command or rule for {prefix}{words[0]}.")
+    else:
+        # Might still be a prefixless rule
+        command_fun, extra_args = _get_rule(words, words_eol, _prefixless_rules)
+        if command_fun:
+            log.debug(f"Prefixless rule {getattr(command_fun, '__name__', '')} matching {words[0]} "
+                      f"found.")
+
+    if command_fun:
+        user = await User.from_whois(bot, sender)
+        context = Context(bot, user, channel, words, words_eol)
+        return await command_fun(context, *extra_args)
+
+
+def _get_rule(words: List[str], words_eol: List[str],
+              from_dict: Dict[Pattern, "_RuleTuple"]) -> Tuple[Optional[Callable], tuple]:
+    """
+    Attempt to find a rule in the given dict of patterns and rules.
+
+    Args:
+        words: Words of the message.
+        words_eol: Words of the message, but each including everything to the end of it.
+        from_dict: Dict mapping patterns to their rules.
+
+    Returns:
+        (Callable or None, tuple):
+            2-tuple of the command function and the extra args that it should
+            be called with.
+    """
+    for pattern, (fun, full_message, pass_match) in from_dict.items():
+        if full_message:
+            match = pattern.match(words_eol[0])
+        else:
+            match = pattern.match(words[0])
+
+        if match is not None:
+            if pass_match:
+                return fun, (match,)
+            else:
+                return fun, ()
+    else:
+        return None, ()
 
-    # check for trigger
-    assert message.startswith(prefix), f"message passed that did not contain prefix."
 
-    log.debug(f"Trigger! {message}")
+def _split_message(string: str) -> Tuple[List[str], List[str]]:
+    """
+    Split up a string into words and words_eol
 
-    # remove command prefix
-    raw_command: str = message.lstrip(prefix)
+    Args:
+        string: Any string.
 
+    Returns:
+        (list of str, list of str):
+            A 2-tuple of (words, words_eol), where words is a list of the words of *string*,
+            seperated by whitespace, and words_eol is a list of the same length, with each element
+            including the word and everything up to the end of *string*
```

Comment:
perhaps an example in the docstring could be valuable since we have begun adopting doctest?

Reply:
In a private member, I wasn't going to bother with extensive documentation. But I guess it can't hurt. Will add.